tools/python/boutiques/localExec.py

```diff
@@ -514,6 +525,21 @@ def _cleanUpAfterSingPull(self, lockDir):
         if "SINGULARITY_PULLFOLDER" in os.environ:
             del os.environ["SINGULARITY_PULLFOLDER"]
 
+    def _isDockerInstalled(self):
+        return not subprocess.Popen("type docker", shell=True).wait()
+
+    # Chooses whether to use Docker or Singularity based on the
+    # descriptor, executor options and if Docker is installed.
+    def _chooseContainerTypeToUse(self, conType, forceSing=False,
+                                  forceDocker=False):
+        if (self._isDockerInstalled() and
+                (conType == 'docker' and not forceSing
+                    or forceDocker)):
+            return "docker"
+        elif (conType == 'singularity' and not forceDocker
```

Comment:
Using `else` would mean that it would always return Singularity if it's not Docker, and then it would never hit the part of `prepare` that throws an error if the container type is neither Singularity not Docker. However, since descriptors would fail validation if the container type isn't Singularity or Docker, we probably don't even need that error.

Reply:
I removed such error code in localExec recently as it was dead code. I don't think we need that error indeed: the code assumes in many places that the descriptor is valid. And if we add another type of containers we'd have to revise it anyway.